components/feral-connectd/cdp.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"sync"
+
+	"github.com/gorilla/websocket"
+)
+
+// CDP Methods
+const (
+	NavigateMethod = "Page.navigate"
+	EvaluateMethod = "Runtime.evaluate"
+)
+
+type CDPClient struct {
+	conn   *websocket.Conn
+	reqID  int
+	config *Config
+	mu     sync.Mutex
+	done   chan struct{}
+}
+
+// NewCDPClient creates a new CDP client
+func NewCDPClient(config *Config) *CDPClient {
+	return &CDPClient{
+		config: config,
+		reqID:  0,
+		done:   make(chan struct{}),
+	}
+}
+
+// InitCDP fetches WS endpoint and dials Chromium
+func (c *CDPClient) InitCDP(ctx context.Context) error {
+	// Fetch JSON with websocket debugger URL
+	resp, err := http.Get(c.config.CDPHost + ":" + strconv.Itoa(c.config.CDPPort) + "/json")
+	if err != nil {
+		return fmt.Errorf("failed to fetch debug targets: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read targets: %w", err)
+	}
+
+	var targets []struct {
+		Type                 string `json:"type"`
+		WebSocketDebuggerURL string `json:"webSocketDebuggerUrl"`
+	}
+	if err := json.Unmarshal(body, &targets); err != nil {
+		return fmt.Errorf("invalid targets format: %w", err)
+	}
+
+	// Connect to CDP websocket
+	for _, t := range targets {
+		if t.Type == "page" {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+
+			c.conn, _, err = websocket.DefaultDialer.Dial(t.WebSocketDebuggerURL, nil)
+			if err != nil {
+				return fmt.Errorf("cdp dial error: %w", err)
+			}
+			log.Println("Connected to Chromium CDP page target:", t.WebSocketDebuggerURL)
+
+			// Start goroutine to handle context cancellation
+			go func() {
```

Comment:
Should be outside of the loop. you could consider to bring this into dedicated function.